torch/csrc/api/src/optim/adam.cpp

```diff
@@ -12,57 +12,171 @@
 
 namespace torch {
 namespace optim {
-AdamOptions::AdamOptions(double learning_rate)
-    : learning_rate_(learning_rate) {}
+AdamOptions::AdamOptions(double lr) : lr_(lr) {}
+
+bool operator==(const AdamOptions& lhs, const AdamOptions& rhs) {
+  return (lhs.lr() == rhs.lr()) &&
+         (lhs.beta1() == rhs.beta1()) &&
+         (lhs.beta2() == rhs.beta2()) &&
+         (lhs.eps() == rhs.eps()) &&
+         (lhs.weight_decay() == rhs.weight_decay() &&
+         (lhs.amsgrad() == rhs.amsgrad()));
+}
+
+void AdamOptions::serialize(torch::serialize::OutputArchive& archive) const {
+  _TORCH_OPTIM_SERIALIZE_TORCH_ARG(lr);
+  _TORCH_OPTIM_SERIALIZE_TORCH_ARG(beta1);
+  _TORCH_OPTIM_SERIALIZE_TORCH_ARG(beta2);
+  _TORCH_OPTIM_SERIALIZE_TORCH_ARG(eps);
+  _TORCH_OPTIM_SERIALIZE_TORCH_ARG(weight_decay);
+  _TORCH_OPTIM_SERIALIZE_TORCH_ARG(amsgrad);
+}
+
+void AdamOptions::serialize(torch::serialize::InputArchive& archive) {
+  _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(double, lr);
+  _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(double, beta1);
+  _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(double, beta2);
+  _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(double, eps);
+  _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(double, weight_decay);
+  _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(bool, amsgrad);
+}
+
+bool operator==(const AdamParamState& lhs, const AdamParamState& rhs) {
+  return (lhs.step() == rhs.step()) &&
+          torch::equal(lhs.exp_avg(), rhs.exp_avg()) &&
+          torch::equal(lhs.exp_avg_sq(), rhs.exp_avg_sq()) &&
+          !(lhs.max_exp_avg_sq().defined() ^ rhs.max_exp_avg_sq().defined());
+}
+
+void AdamParamState::serialize(torch::serialize::OutputArchive& archive) const {
+  _TORCH_OPTIM_SERIALIZE_TORCH_ARG(step);
+  _TORCH_OPTIM_SERIALIZE_TORCH_ARG(exp_avg);
+  _TORCH_OPTIM_SERIALIZE_TORCH_ARG(exp_avg_sq);
+  if(max_exp_avg_sq().defined()) {
```

Comment:
I think we should move the logic for skipping undefined tensor into `_TORCH_OPTIM_SERIALIZE_TORCH_ARG`, similar to what we do for `_TORCH_OPTIM_DESERIALIZE_TORCH_ARG`.